src/compiler/performance.ts

```diff
@@ -1,146 +1,209 @@
+        const durations = new Map<string, number>();
+        const durationMarks = new Set<string>();
+        let statistics: ESMap<string, Statistic> | undefined;
+
         return {
-            enter,
-            exit
+            createTimerIf,
+            createTimer,
+            mark,
+            measure,
+            addStatistics,
+            getCount,
+            getDuration,
+            forEachMeasure,
+            forEachCount,
+            forEachStatistics,
+            isEnabled,
+            enable,
+            disable,
         };
 
-        function enter() {
-            if (++enterCount === 1) {
-                mark(startMarkName);
+        function createTimerIf(condition: boolean, measureName: string, startMarkName: string, endMarkName: string) {
+            return condition ? createTimer(measureName, startMarkName, endMarkName) : nullTimer;
+        }
+
+        function createTimer(measureName: string, startMarkName: string, endMarkName: string): Timer {
+            let enterCount = 0;
+            return {
+                enter,
+                exit
+            };
+
+            function enter() {
+                if (++enterCount === 1) {
+                    mark(startMarkName);
+                }
+            }
+
+            function exit() {
+                if (--enterCount === 0) {
+                    mark(endMarkName);
+                    measure(measureName, startMarkName, endMarkName);
+                }
+                else if (enterCount < 0) {
+                    Debug.fail("enter/exit count does not match.");
+                }
             }
         }
 
-        function exit() {
-            if (--enterCount === 0) {
-                mark(endMarkName);
-                measure(measureName, startMarkName, endMarkName);
+        /**
+         * Marks a performance event.
+         *
+         * @param markName The name of the mark.
+         */
+        function mark(markName: string) {
+            if (enabled) {
+                const count = counts.get(markName) ?? 0;
+                counts.set(markName, count + 1);
+                marks.set(markName, timestamp());
+                performanceImpl?.mark(markName);
             }
-            else if (enterCount < 0) {
-                Debug.fail("enter/exit count does not match.");
+        }
+
+        /**
+         * Adds a performance measurement with the specified name.
+         *
+         * @param measureName The name of the performance measurement.
+         * @param startMarkName The name of the starting mark. If not supplied, the point at which the
+         *      profiler was enabled is used.
+         * @param endMarkName The name of the ending mark. If not supplied, the current timestamp is
+         *      used.
+         */
+        function measure(measureName: string, startMarkName: string, endMarkName: string) {
+            if (enabled) {
+                durationMarks.add(startMarkName).add(endMarkName);

```

Comment:
Why are these excluded here? They wouldn't be excluded by the native User Timings API.